### ISSUE TITLE
NUnit Console Runner and Visual Studio Runner ignore parallize settings.

### ISSUE THREAD

**Issue by Mrnikbobjeff** [nunit/nunit#3438]
I have a single test assembly containing 100 unit tests. As some of them are integration tests which depend on external (mutable) state I require the tests to be executed in a serial manner. To accomplish this I started of with the usual .runsettings file and added the following two tags:
```
<DisableParallelization>true</DisableParallelization>
<MaxCpuCount>1</MaxCpuCount>
```

This did not lead to the desired outcome, so I applied `[Parallelizable(ParallelScope.None)]` to my fixture, and to be extra paranoid to all tests in the fixture. None of these changes lead to tests not being executed executed in parallel. I checked #3371 as well as that seemed to relate to my issue, but to no avail. The behaviour is reproducible in the console runner as well, setting --workers=1 --process=Single in combination with the attribute leads to the same result as the visual studio runner. I would be grateful for any insight into this mess.

**Comment by CharliePoole** [comment]
I added a backtick fence around your XML settings, which were not visible in the post.

A few observations, followed by some quesetions:
:
 * The `<DisableParallelization> setting has to do with (not) running multiple test assemblies in parallel. If you only have one test assembly, then it has no effect.
 * The `<MaxCpuCount>` setting has no effect if you have disapbled parallelization by Visual Studio.
 * Both of the above settings are used __only__ by VS for it's own parallelization and have nothing to do with NUnit's parallelization.
 * NUnit (you probably know) has two types of parallelization: within an assembly and between assemblies. Since you appear to have only one test assembly, only the second type is relevant to you.
 * Issue #3371 relates to a very specific situation involving parameterized test cases. If you don't have that situation, it's not relevant to you.
 * The nunit console runner `--process` option has to do with where your test assemblies are run. In particular, `--process=Single` is a synonym for `--inprocess` and causes the tests to run in the same process as the console runner itself, rather than in a separate process. It doesn't have any impact on parallelization anyway, since you only have one assembly.
 * The `--workers` option provides a way to set the number of parallel workers __within an assembly__ and is the closest to what you are looking for. However, it's main use is to __override__ the normal behavior of your tests, which is determined by the use of `ParallelizableAttribute` and `NonParallelizableAttribute` in the code itself. If your tests have to __always__ run in a certain way, it's better to use the attriutes appropriately. Otherwise, someone else might run the tests with the wrong options and be confused by the results.

After the observations, some questions:

 * Do you have other test fixtures, which run in parallel, in the same assembly? If __none__ of your tests should run in parallel, you should simply omit all use of `[Parallelizable]`. If you never use the attribute, NUnit will not run tests in parallel at all.
 * How are you determining that the tests run in parallel? It could be that they are running sequentially but that some tests are not cleaning up correctly after themselves.
 * If you can create a minimual example, it may help to diagnose the problem.

**Comment by Mrnikbobjeff** [comment]
@CharliePoole I thank you for your detailed answer, I definitely learned things I did not know about parallelization in Nunit. I am using the visual studio test runner, and I ensure the validity of the tests by running them on their own, executing only the test which is failing due to the incorrect shared state. If I do this for every test, no test will fail. I will try to get a version running which simply prints statements once the tests starts to ensure that they are actually running in parallel. Until now I just assumed they did because of the test runner UI which flags tests as failing or succeeding at the same time. The exception in failed tests relays simultaneous access to a shared resource, which strengthened my suspicion. I did not use the Parallize Attribute at all. I actually did switch to Xunit for this test assembly as I knew about their parallelization technique due to fixing a bug there a while ago, which also alleviated my concern. Nevertheless I will try to get a solution which reproduces the issue.
